### PR TITLE
Require cluster id for AWS

### PR DESCRIPTION
**Pull request by rrati**
Backporting of:

https://github.com/kubernetes/kubernetes/pull/48612
https://github.com/kubernetes/kubernetes/pull/49215

Plus changes in master-controller startup to look for the flag to allow untagged clouds or exit.

### CODE REVIEW

**Comment by mfojtik** [comment]
@rrati are my hopes too high thinking that this might fix https://bugzilla.redhat.com/show_bug.cgi?id=1491202 ?

File: vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/gce/gce.go [hunk 1]
```diff
@@ -362,6 +362,11 @@ func (gce *GCECloud) ScrubDNS(nameservers, searches []string) (nsOut, srchOut []
 	return nameservers, srchOut
 }
 
+// HasClusterID returns true if the cluster has a clusterID
+func (gce *GCECloud) HasClusterID() bool {
+	return true
```
Comment (mfojtik) [inline vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/gce/gce.go]: @rrati isn't GCE cloud using Initialize() to get the cluster id and this should return true only when the clusterID is really available?

Comment (rrati) [reply vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/gce/gce.go]: @mfojtik Yes, this really should be checking if the GCE Cluster ID is set or not.  I'll log an issue upstream and correct it.

**Comment by rrati** [comment]
@mfojtik I don't see this fixing https://bugzilla.redhat.com/show_bug.cgi?id=1491202 .  In this case, the controller-manager would exit if the cloud provider isn't labelled unless they pass an argument to allow an untagged cluster.  It won't do anything to actually set the ClusterID.

File: vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/gce/gce.go (hunk 1 above)
Comment (rrati) [reply vendor/k8s.io/kubernetes/pkg/cloudprovider/providers/gce/gce.go]: @mfojtik Actually, looking at the usage of GCE ClusterID this function should be as is, ie always returning true.

**Comment by rrati** [comment]
@mfojtik I think this needs a lgtm label if it is approved for merging

**Comment by enj** [comment]
@openshift/api-review so this took down a cluster by forcing the controller processes into a crash loop.  This requires further discussion to decide if this needs to be the default.

@rrati @mfojtik

**Comment by deads2k** [comment]
> @openshift/api-review so this took down a cluster by forcing the controller processes into a crash loop. This requires further discussion to decide if this needs to be the default.

Rather than trying to compensate for this in the master-config.yaml (where user set options and auto-set options will collide), if we know that we're going to need this information and we know that we'll never be able to auto-create this information, we should block the upgrade until the information is present.  Doing this is just going to lead to confusion later about whether we can or can't automatically remove it after doing the same upgrade block in 3.8.

I'd actually be ok pairing a still-broken default (for one release!), with blocking on upgrade.

**Comment by enj** [comment]
@sdodson @jupierce can you guys sync with @deads2k to add this blocking pre-upgrade check?

**Comment by sdodson** [comment]
@enj https://bugzilla.redhat.com/show_bug.cgi?id=1491399

**Pull request merged**